src/preprocessing/passes/bv_ackermann.cpp

```diff
@@ -41,55 +40,150 @@ namespace passes {
-    FunctionToArgsMap& fun_to_args,
-    SubstitutionMap& fun_to_skolem)
+void addLemmaForPair(TNode args1,
+                     TNode args2,
+                     const TNode func,
+                     AssertionPipeline* assertionsToPreprocess,
+                     NodeManager* nm)
+{
+  Node args_eq;
+
+  if (args1.getKind() == kind::APPLY_UF)
+  {
+    Assert(args1.getOperator() == func);
+    Assert(args2.getKind() == kind::APPLY_UF
+                 && args2.getOperator() == func);
+    Assert(args1.getNumChildren() == args2.getNumChildren());
+
+    std::vector<Node> eqs(args1.getNumChildren());
+
+    for (unsigned i = 0, n = args1.getNumChildren(); i < n; ++i)
+    {
+      eqs[i] = nm->mkNode(kind::EQUAL, args1[i], args2[i]);
+    }
+    args_eq = bv::utils::mkAnd(eqs);
+  }
+  else
+  {
+    Assert(args1.getKind() == kind::SELECT && args1[0] == func);
+    Assert(args2.getKind() == kind::SELECT && args2[0] == func);
+    Assert(args1.getNumChildren() == 2);
+    Assert(args2.getNumChildren() == 2);
+    args_eq = nm->mkNode(kind::EQUAL, args1[1], args2[1]);
+  }
+  Node func_eq = nm->mkNode(kind::EQUAL, args1, args2);
+  Node lemma = nm->mkNode(kind::IMPLIES, args_eq, func_eq);
+  assertionsToPreprocess->push_back(lemma);
+}
+
+void addChildrenToStack(TNode node, std::stack<TNode>* stack)
+{
+  for (const TNode& arg: node)
+  {
+    stack->push(arg);
+  }
+}
+
+void storeFunctionAndAddLemmas(TNode func,
+                               TNode term,
+                               FunctionToArgsMap& fun_to_args,
+                               SubstitutionMap& fun_to_skolem,
+                               AssertionPipeline* assertions,
+                               NodeManager* nm,
+                               std::stack<TNode>* stack)
 {
   if (fun_to_args.find(func) == fun_to_args.end())
   {
-    fun_to_args.insert(make_pair(func, NodeSet()));
+    fun_to_args.insert(make_pair(func, TNodeSet()));
   }
-  NodeSet& set = fun_to_args[func];
+  TNodeSet& set = fun_to_args[func];
   if (set.find(term) == set.end())
   {
-    set.insert(term);
     TypeNode tn = term.getType();
-    Node skolem = NodeManager::currentNM()->mkSkolem(
+    Node skolem = nm->mkSkolem(
         "BVSKOLEM$$",
         tn,
         "is a variable created by the ackermannization "
         "preprocessing pass for theory BV");
+    for (const auto& t : set)
+    {
+      addLemmaForPair(t, term, func, assertions, nm);
+    }
     fun_to_skolem.addSubstitution(term, skolem);
+    set.insert(term);
+    /* Add the arguments of term (newest element in set) to the stack, so that
+     * collectFunctionsAndLemmas will process them as well. 
+     * This is only needed if the set has at least two elements 
+     * (otherwise, no lemma is generated).
+     * Therefore, we defer this for term in case it is the first element in the set*/
+    if (set.size() == 2) 
+    {
+      for (const TNode& elem : set)
+      {
+        addChildrenToStack(elem, stack);
```

Comment:
With std::vector as stack you can use a one liner: `stack.insert(stack.end(), elem.begin(), elem.end());`